crates/cubecl-linalg/src/matmul/tests/test_utils.rs

```diff
@@ -196,6 +343,91 @@ where
     out
 }
 
+pub(crate) fn matmul_cpu_reference_quantized(
+    lhs: &[u8],
+    rhs: &[u8],
+    lhs_zero_offset: i32,
+    rhs_zero_offset: i32,
+    problem: &MatmulProblem,
+) -> Vec<i32>
+where
+{
+    let m = problem.m;
+    let n = problem.n;
+    let k = problem.k;
+    let num_batches = problem.num_batches();
+
+    let (b_lhs, b_rhs) = problem.batches.clone();
+    assert!(b_lhs.len() == b_rhs.len(), "Cpu reference only works with batches of equal length. Please pad the shortest one with ones at the beginning.");
+
+    let lhs_strides = strides(problem, Ident::Lhs);
+    let rhs_strides = strides(problem, Ident::Rhs);
+    let out_strides = strides(problem, Ident::Out);
+
+    let mut out = vec![0; m * n * num_batches];
+
+    for nth_batch in 0..num_batches {
+        let batch_out = nth_batch * m * n;
+        let mut batch_lhs = 0;
+        let mut batch_rhs = 0;
+        for b in 0..b_lhs.len() {
+            let tmp = batch_out / out_strides[b];
+            batch_lhs += tmp % b_lhs[b] * lhs_strides[b];
+            batch_rhs += tmp % b_rhs[b] * rhs_strides[b];
+        }
+
+        // Perform matmul
+        for row in 0..m {
+            for col in 0..n {
+                for middle in 0..k {
+                    let lhs_index = row * k + middle;
+                    let rhs_index = middle * n + col;
+                    let out_index = row * n + col;
+
+                    let l = lhs[batch_lhs + lhs_index] as u16;
+                    let r = rhs[batch_rhs + rhs_index] as u16;
+                    let prod = l * r;
+
+                    out[batch_out + out_index] += prod as i32;
+                }
+            }
+        }
+
+        // Substract rhs_zero_offset * sum_rows(lhs)
+        for row in 0..m {
+            let mut sum = 0;
+            for col in 0..k {
+                sum += lhs[batch_lhs + row * k + col] as i32;
+            }
+            sum *= rhs_zero_offset;
+            for col in 0..n {
+                out[batch_out + row * n + col] -= sum;
+            }
+        }
+
+        // Substract lhs_zero_offset * sum_cols(rhs)
+        for col in 0..n {
+            let mut sum = 0;
+            for row in 0..k {
+                sum += rhs[batch_lhs + row * n + col] as i32;
+            }
+            sum *= lhs_zero_offset;
+            for row in 0..m {
+                out[batch_out + row * n + col] -= sum;
+            }
+        }
+
+        // Add final constant term
+        for row in 0..m {
+            for col in 0..n {
+                out[batch_out + row * n + col] += (k as i32) * lhs_zero_offset * rhs_zero_offset;
+            }
+        }
```

Comment:
Here you can see the flow for the quantized matmul.

Reply:
Another way to test it is to dequantize a tensor, perform matmul, requantize the tensor and see if the quantize matmul works.